Show device mode

andrzejtp opened the pull request:
This is an attempt to address https://github.com/boardswarm/boardswarm/issues/14

There's an "Initial" state of the device. Maybe we should instead somehow query the device for its initial state. 

Perhaps we could also want to show device's `id` and/or `boardswarm.name`.

sjoerdsimons commented on boardswarm-cli/src/ui.rs:
```diff
@@ -223,23 +259,35 @@ pub async fn run_ui(
                     match i {
                         Input::PowerOn => {
-                            device.change_mode("on").await.unwrap();
+                            let _ = device.change_mode("on").await;
```
while just unwrapping isn't great; hiding the failure isn't good either :)

sjoerdsimons commented on boardswarm-cli/src/ui_term.rs:
```diff
@@ -5,13 +5,31 @@ use ratatui::{
 };
 
+#[derive(Clone, Copy, Debug)]
+pub enum TerminalState {
```
These are not states; just device modes and device modes are not enums they're aribitrary strings (on/off are just conventions); Reset is not a state either and neither is initial.

andrzejtp replied:
OTOH if I want to display different modes with different attributes (e.g. on would be black on green, while off would be black on red), then I need to start interpreting the string, which seems not the greatest option. What's more, future localization might mean that actual strings passed are unknown to the programmer.

sjoerdsimons commented on boardswarm-cli/src/ui_term.rs:
```diff
@@ -56,6 +74,20 @@ impl Widget for UiTerm<'_> {
+                TerminalState::Initial => style.bg(Color::Gray),
+            };
+            let span = Span::styled(str, style);
+            buf.set_span(area.width - width, area.height - 1, &span, width);
```
this probably should not be in the uiterm widget but instead in e.g. a status bar widget

sjoerdsimons commented on boardswarm-cli/src/ui.rs:
```diff
@@ -206,7 +238,11 @@ pub async fn run_ui(
                         Input::Down => { terminal.scroll_down(); },
                         Input::ScrollReset => { terminal.scroll_reset(); },
-                        _ => (),
+                        Input::PowerOn => { terminal.set_power_on(); },
+                        Input::PowerOff => { terminal.set_power_off(); },
+                        Input::PowerReset => { terminal.set_power_reset(); },
+                        Input::ShowState => { terminal.show_state(); },
```
Really the device mode should be coming from the actual device mode; not what a ui has sent ..  You can listen to the device-info stream ; See how e.g. the implementation of `boardswarm-cli device X info --follow`

andrzejtp replied:
Would the right place to listen to device info stream be https://github.com/boardswarm/boardswarm/blob/7c04a7c8d41ba0fbfdb2892926157c7c93a3b1f1/boardswarm-cli/src/ui.rs#L196?

sjoerdsimons replied:
I'd probably do it in a seperate task and only message the main thread when there is a mode update specificially such that it can trigger a next frame.

andrzejtp commented:
Added layout and a separate widget for device mode

andrzejtp commented on boardswarm-cli/README.md:
```diff
@@ -31,6 +31,7 @@ The UI has the follow keyboard shortcuts:
 * ^a o: Change the device to mode "on"
 * ^a f: Change the device to mode "off"
 * ^a r: Reset the device's power (same as changing mode to "off" then "on")
+* ^a i: Show device's current mode
```
Have you tried actually running the code?

The idea is that the mode _is_ displayed automatically on any change. But then after the user starts typing the mode becomes _not displayed_ in order to not clutter the terminal. Should the user want to recall what the current mode is, they can issue ctrl-a i - the mode is then shown and is displayed until the user starts typing again.

obbardc replied:
No this is purely a review at this point ;-)

i wonder why it makes sense to hide it ? possibly we could always keep a row at the top/bottom of the screen for the boardswarm ui (with the console taking up -2 rows)

andrzejtp replied:
For now the terminal is fixed at 80x24. Permanently stealing one or even two lines from that seems a crime to me.

sjoerdsimons replied:
We should indeed never steal lines; A permanent status bar is a nice to have but it should never be prioritised over the terminal

adalessandro replied:
Is there a particular reason why the area is fixed to 80x24?

sjoerdsimons replied:
@adalessandro  see https://github.com/boardswarm/boardswarm/pull/153#pullrequestreview-2560369986

andrzejtp commented on boardswarm-cli/src/ui.rs:
```diff
@@ -206,14 +269,23 @@ pub async fn run_ui(
+                }
+                Some(new_state) = state_rx.recv() => {
+                    if new_state == "on" {
+                        terminal.set_power_on("PowerOn".to_string());
```
I think it makes sense to decouple the mode from definition of how mode is displayed. By that I mean that if e.g. "off" is displayed as black on red, maybe other kinds of modes will also want to be displayed black on red, so I called this kind of display "NonFunctional", as opposed to "Functional" which is black on green. That is why the modes are not relayed 1:1 to the ui. On top of that, the string needs to be parsed to know how to display it. Not great maybe. And then if localization (i.e. displaying in national languages rather than English) is added, then the strings passed will not be known up front to the programmer, so the programmer will effectively be unable to parse (in a sane way) even if they wanted to.

For now this question boils down to "where should the mode string be parsed in order to decide how the mode is displayed?". Any preference/rationale for that?

andrzejtp commented on boardswarm-cli/src/ui.rs:
```diff
@@ -235,17 +307,38 @@ pub async fn run_ui(
+                        Input::Bytes(data) => {
+                            // send an empty "Bytes" to notify that some
+                            // ordinary characters have been entered
+                            input_tx.send(Input::Bytes(Bytes::new())).await.unwrap();
```
This is directly related to mode ceasing being displayed as soon as the user starts typing.

obbardc replied:
oh that really wasn't clear :-) perhaps you could add a comment about what is going on ?

andrzejtp commented on boardswarm-client/src/device.rs:
```diff
@@ -232,6 +232,14 @@ impl Device {
         Ok(())
     }
 
+    pub async fn device_info(
```
:+1:

sjoerdsimons replied:
the Device already monitors the info stream; so with this api you're now receiving two info streams;  The Device struct was really meant for high-level api's, so it's a bit odd to directly expose the low-level protocol like that again

sjoerdsimons replied:
fwiw looking overall; it may make sense to push watching the device mode into the Device itself; either with a one-shot api or by exposing a stream of mode-changes. Should probably also add an api to get the current mode

andrzejtp commented:
@obbardc This time I have simplified the code a bit. The status is displayed always, which uses the line in the bottom of the display. That's not ideal, because we're left with just 23 lines, but OTOH relaxing 80x24 is probably an orthogonal problem.

obbardc commented on boardswarm-cli/src/ui.rs:
```diff
@@ -209,11 +259,19 @@ pub async fn run_ui(
                         _ => (),
                     }
                 }
+                Some(new_state) = state_rx.recv() => {
```
how about other states ? can we just pass them straight through to the UI ?

a match statememnt may be cleaner here

andrzejtp commented:
@obbardc In this iteration on top of addressing your comment about matching the state string, I've addressed https://github.com/boardswarm/boardswarm/issues/13 in such a way that we don't hang but instead the message is relayed to the message area which I've split from the display bottom line.

andrzejtp commented:
@obbardc I mistakenly pushed an old branch and corrected it by pushing the latest changes. This time https://github.com/boardswarm/boardswarm/issues/13 is also addressed by relaying the message to the user instead of hanging at unwrap().

sjoerdsimons commented on boardswarm-cli/src/ui.rs:
```diff
@@ -203,17 +289,30 @@ pub async fn run_ui(
+                    let state: String = new_state;
+                    match state.as_str() {
+                        "on" => terminal.set_functioning("ON".to_string()),
+                        "off" => terminal.set_non_functioning("OFF".to_string()),
```
these are devie modes not terminal modes though; functioning/non-functioning is weird jargon as welll ;  why not just on/off/other(string) ?

andrzejtp replied:
I'm okay with renaming `TerminalMode` to `DeviceMode`.

Functioning/NonFunctioning: Perhaps there will be states which want to be displayed the same way "off" or "on" is displayed. Let's imagine there will appear a state called "relay" which wants to be displayed black on green. Then the match arm becomes:

```
"relay" => terminal.set_functioning("RELAY".to_string()),
```

Otherwise it will have to be:

```
"relay" => terminal.set_on("RELAY".to_string()),
```

sjoerdsimons replied:
can you please not yell in strings ;) Yes there might be cases we may want to do differently (or maybe we could stick a hint in the protocol for broad categories), but regardless we can simply add more "known" items in the enum. 

Fwiw it would also be nicer to implement From<&str> for your mode enum

sjoerdsimons commented on boardswarm-cli/src/ui.rs:
```diff
@@ -50,21 +70,79 @@ impl Terminal {
-                let term_area = Rect::new(0, 0, 80.min(size.width), 24.min(size.height));
-                f.render_widget(term, term_area);
+                let term_area = Rect::new(0, 0, COLUMNS.min(size.width), ROWS.min(size.height));
+                let area_chunks = Layout::default()
```
fwiw you can use `.areas` and destructure instead of having to do array indexes

sjoerdsimons commented on boardswarm-cli/src/ui.rs:
```diff
@@ -50,21 +70,79 @@ impl Terminal {
+                let term_area = Rect::new(0, 0, COLUMNS.min(size.width), ROWS.min(size.height));
+                let area_chunks = Layout::default()
+                    .direction(Direction::Vertical)
+                    .constraints([Constraint::Min(1), Constraint::Length(1)])
```
I assume this means that if the terminal size is actually 80x24 you'll only see 23 lines? Would it make sense to add a min constraint of 24 for the term so we prioritize over showing that fully *over* showing a status bar?

andrzejtp replied:
Yes, in the current shape of the series I do steal the lowest line for the status bar.

You guys settle on what you want first https://github.com/boardswarm/boardswarm/pull/90#discussion_r1622021635

Actually I imagine that you maybe _don't_ know and this thread is exactly about _learning_ what we want and I'm okay with that. Only I wouldn't like to go in circles too many times. ;P

It seems to me that relaxing the 80x24 is a bit of an orthogonal problem. IOW trying to conditionally display the status bar (sometimes display, sometimes not) is effectively trying to work around not being able to use other terminal sizes. Then there's a question of _when_ to display and when not. Then the user might want to recall what the current mode is. It's already been tried in this thread. Displaying the status bar all the time does simplify the code.

sjoerdsimons replied:
I don't need to settle on anything really :) Displaying the console is the obvious primary function of the UI everything else is secondary (hence that's all it does at the moment); We shouldn't ever force the terminal area to be smaller then what it wants to be as that will likely hide output lines; e.g. if the last line is hidden your shell input will disappear quite quickly; if the top line is hidden you're missing an important part of e.g. `top`.

Not being able to set the size is orthogonal to this and primarily it's not really a boardswarm client call but depends on what the remote console side thinks it's rendering to, so we can't just decide "relax" and change size because we want to fit a statusbar :) Imho this should go as far as the UI displaying an error if the output size is smaller then its terminal size, because, really, we should never hide things in the terminal as that quickly ends up confusing. 

For 80x24, that's still the default when opening most terminal emulators these days even if it's somewhat historical. And similar a serial tty on linux/u-boot serial output etc all also start at 80x24, hence the hardcoding in the UI at the moment. But that also fundamentally means we should support displaying 80x24  on a 80x24, hence dropping out the status bar in those cases to not be annoying in that situation :)

sjoerdsimons commented on boardswarm-cli/src/ui.rs:
```diff
@@ -246,6 +350,18 @@ pub async fn run_ui(
 
+    let _device_state_watcher = tokio::spawn(async move {
+        let mut d = dev_clone.device_info().await.unwrap();
+        let mut current_mode = None;
```
The device mode already knows the current mode as it will wait for a first info on creation.
